Re-Work Session Storage

dwyeradam opened the pull request:
Store session data in the server side session. The only data that should be in the client side cookie is enough information to identify the user.

jonnyshaw89 commented on app/authentication/authenticator.py:
```diff
@@ -76,6 +77,17 @@ def store_session(metadata):
+        'ru_name': metadata.get('ru_name'),
+        'period_str': metadata.get('period_str')
+    }
+    return json.dumps(session_data)
```
I think we should be consistent in the `dumps` and `loads` we pass in json to the `SessionStore` but return a dict. I think we should pass in the Dict here and do all of the serialisation inside `SessionStore`

jonnyshaw89 commented on app/authentication/session_storage.py:
```diff
@@ -18,47 +19,56 @@ class SessionStorage:
-                logger.debug("eq_session_id from user's cookie not found in database")
-        else:
-            logger.debug("eq_session_id does not exist in user's cookie")
+        if eq_session is None:
```
Why have we changed this behaviour? when would we try to delete a session we don't know exists?

dwyeradam replied:
I need to change this. It's actually to fix a test so I need to rewrite the test in test_database_session_manager.py.

jonnyshaw89 commented on app/authentication/session_storage.py:
```diff
@@ -18,47 +19,56 @@ class SessionStorage:
 
     @staticmethod
-    def _get_user_session(eq_session_id):
+    def _get_user_session():
```
Should we be caching this result?
How many times in a request will this be hit?

dwyeradam replied:
I did originally cache it however it's only called once per request so not really required. By not caching it will make it easier to use session storage for dynamic values in the future.

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -40,21 +41,26 @@
     metadata = get_metadata(current_user)
-    if metadata:
-        logger.bind(tx_id=metadata['tx_id'])
+    if not bool(metadata):
```
Why do we need to cast this to a bool? can't we check not None?

jonnyshaw89 replied:
Or even, why would it ever be None?

dwyeradam replied:
This is after submission when trying to get back into the questionnaire. get_metadata() returns an empty dict as the user is now still logged in but the metadata has been cleared. I'll look at the logic around returning none instead of an empty dict.

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -170,13 +176,15 @@ def post_household_composition(routing_path, **kwargs):
 
-    if survey_completed_metadata:
-        metadata_context = build_metadata_context_for_survey_completed(survey_completed_metadata)
+    if not bool(metadata):
```
Can't you just check for `None`?

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -273,19 +283,10 @@ def submit_answers(routing_path, eq_id, form_type, collection_id):
-
         get_questionnaire_store(current_user.user_id, current_user.user_ik).delete()
-        remove_survey_session_data()
+        # remove_survey_session_data()
```
Remove comment

jonnyshaw89 commented on app/authentication/session_storage.py:
```diff
@@ -69,7 +78,7 @@ def _get_user_session(eq_session_id):
                          session_id=eq_session.eq_session_id,
                          user_id=eq_session.user_id,
-                         timestamp=eq_session.timestamp.isoformat())
+                         created_at=eq_session.created_at.isoformat())
```
Do we need to log the created time?

dwyeradam replied:
We had it before (timestamp) so i've left it in. We probably don't though so worth discussing with @ajmaddaford.

jonnyshaw89 replied:
With this in it errors when loading a session if that session was started on the previous version of code

jonnyshaw89 commented on app/data_model/database.py:
```diff
@@ -32,15 +32,17 @@ class EQSession(base):
 
     def __repr__(self):
-        return "<EQSession('%s', '%s', '%s')>" % (self.eq_session_id, self.user_id, self.timestamp)
+        return "<EQSession('%s', '%s', '%s')>" % (self.eq_session_id, self.user_id, self.session_data)
```
Is it valuable to have the data in this?

dwyeradam replied:
Guess not. Will remove.

jonnyshaw89 commented on app/globals.py:
```diff
@@ -24,7 +24,10 @@ def get_metadata(user):
 
     questionnaire_store = get_questionnaire_store(user.user_id, user.user_ik)
-    return questionnaire_store.metadata
+    if questionnaire_store.metadata:
```
If this is false then `questionnaire_store.metadata` would be None? therefore can you not just return `questionnaire_store.metadata` as it was before?

dwyeradam replied:
It will always exist but will be empty after submission. This logic stops it returning an empty dictionary. Will change the later logic to cope with the empty dictionary where expected.

jonnyshaw89 replied:
Yeh if this is the only place we set this to `None` then the other code should just handle the empty dict.

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -40,21 +41,26 @@
                                     import_name=__name__,
                                     url_prefix='/questionnaire/<eq_id>/<form_type>/<collection_id>/')
 
+after_submission_blueprint = Blueprint(name='after_submission',
```
after submission or post submission? just my opinion, ignore it if you like

dwyeradam replied:
I'll toss a coin but I might weight it in your favour as I probably agree.

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -40,21 +41,26 @@
-    if metadata:
-        logger.bind(tx_id=metadata['tx_id'])
+    if metadata is None:
+        raise NoTokenException(401)
```
I do wonder if we are trying to re-implement authentication logic.

dwyeradam replied:
We are really as at this point the user is still authenticated but we don't want to allow them back into the questionnaire. It may be worth creating a new error for this.

jonnyshaw89 replied:
Do we have the submitted flag/time in the session? if so can we check this and redirect them to the thankyou?

dwyeradam replied:
This matches current behaviour if a user attempts to get back into the questionnaire after submission. Another one to discuss with @ajmaddaford. Argument for both but confusion could arise if the user is pressing back but just keeps getting routed to where they were in the first place without any kind of notification.

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -170,13 +176,15 @@ def post_household_composition(routing_path, **kwargs):
-    survey_completed_metadata = current_app.eq['session_storage'].get_survey_completed_metadata()
-    schema = load_schema_from_params(eq_id, form_type)
+@after_submission_blueprint.route('/thank-you', methods=['GET'])
+def get_thank_you():  # pylint: disable=unused-argument
```
you can remove the pylint exclusion as there are no unused arguments on this funciton

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -170,13 +176,15 @@ def post_household_composition(routing_path, **kwargs):
+def get_thank_you():  # pylint: disable=unused-argument
+    session_data = current_app.eq['session_storage'].get_session_data()
+    eq_id, form_type = session_data.get('schema_identifier').split('_', maxsplit=1)
+    metadata = get_metadata(current_user)
```
Here we call `current_app.eq['session_storage'].get_session_data()` which queries `EQSession` and then 2 lines later we call `get_metadata(current_user)` which queries `EQSession` again.
Is there a way we can limit this to 1 call per request?

shahi645 commented on app/authentication/session_storage.py:
```diff
@@ -18,13 +19,13 @@ class SessionStorage:
         self._database = database
 
-    def store_user_id(self, user_id):
+    def store_user_id(self, user_id, session_data):
```
should we rename this method? 🤷‍♂️

ajmaddaford commented on app/authentication/authenticator.py:
```diff
@@ -76,6 +76,17 @@ def store_session(metadata):
+def get_session_data(metadata):
+    session_data = {
+        'tx_id': metadata.get('tx_id'),
+        'schema_identifier': metadata.get('eq_id') + '_' + metadata.get('form_type'),
```
Why not just store `eq_id` and `form_type` in the session? The only place I can see `schema_identifier` used these values are split on the `_`.

jonnyshaw89 replied:
If we only store `eq_id` and `form_type` then loading from url would break?

ajmaddaford replied:
We would store that under a survey_url property, the same as it is in the metadata. Even if a survey is loaded from a url it still has an eq_id and form_type.

ajmaddaford commented on app/authentication/authenticator.py:
```diff
@@ -76,6 +76,17 @@ def store_session(metadata):
+        'ru_name': metadata.get('ru_name'),
+        'period_str': metadata.get('period_str')
+    }
+    return session_data
```
- I think the definition of the session data should be with the session storage rather than authenticator. 
- We should consider making the session data a class so that we have a clearly defined thing that will be returned. It would also mean we could access session properties with `session.property`. Authenticator could create a SessionData class and call a `populate_from_metadata` method.

ajmaddaford reviewed (changes requested):
There is a card in our technical backlog to separate the cookie session out of session_storage, as it's confusing accessing the cookie session while also interacting with the database session. I think we should take this opportunity to do this work. 

We should be able to limit our use of the cookie session to authenticator.py and just pass the session id to session storage. One way we could make it clearer which session is which in authenticator.py would be to do something like `import session as cookie_session`.

jonnyshaw89 commented on app/authentication/session_storage.py:
```diff
@@ -15,52 +18,71 @@
         db.session.commit()
 
+    @staticmethod
+    def _create_session_data_from_dict(data_dict):
```
should this be `from_metadata`? as its not just a dict its specifically the metadata?

dwyeradam replied:
Removed this method. Instead of a second call to this I'm using the object_hook in json.loads to directly get the SessionData object.

jonnyshaw89 commented on app/authentication/authenticator.py:
```diff
@@ -91,3 +93,40 @@ def decrypt_token(encrypted_token):
+    """
+    Removes a user's ik from the session
+    """
+
```
remove this line?

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -170,13 +176,14 @@ def post_household_composition(routing_path, **kwargs):
-    if survey_completed_metadata:
-        metadata_context = build_metadata_context_for_survey_completed(survey_completed_metadata)
+    if not metadata:
+        schema = load_schema_from_params(session_data.eq_id, session_data.form_type)
```
If the version was updated while the user was completing the questionnaire, when they reach the thank-you page it will error.
We should handle this somehow. maybe populate the session_data on submit if it isn't present

dwyeradam replied:
I'll remove the created_at from the debug message as not really needed. I'll need to create an update for session_data that can be checked at submit.

jonnyshaw89 commented on app/authentication/session_storage.py:
```diff
@@ -1,144 +1,98 @@
-        session[EQ_SESSION_ID] = eq_session_id
-        eq_session = EQSession(eq_session_id, user_id)
+        session_data = self._create_session_data_from_meta(metadata)
+        eq_session = EQSession(eq_session_id, user_id, json.dumps(session_data.__dict__))
```
i think the best practice is not to use `__dict__` but to use `vars(session_data)` instead

jonnyshaw89 commented on app/authentication/session_storage.py:
```diff
@@ -1,144 +1,98 @@
+        eq_session = self._get_user_session()
+        if eq_session:
+            session_data = self._create_session_data_from_meta(metadata)
+            eq_session.session_data = json.dumps(session_data.__dict__)
```
same as above

ajmaddaford commented:
From discussions with  @dwyeradam and @jonnyshaw89 we're going to:
- Remove cookie helper
  It's not very useful having a helper class that only has methods for manipulating the cookie dict, it's also not clear where these methods are called that they're dealing with the cookie session.
- Use flask logout hook so we can remove session helper
  Session helper only has one method, it looks like we should be able to hook into the flask-login logout handling to do the same thing.
- Make session a global and migrate SessionStorage to SessionStore (a QuestionnaireStore like way of interacting with the current user session)
  Currently every time the session is accessed it is loaded from the database. It would be simpler and more efficient to load this once with each request in a similar way to questionnaire store. 
- Write session data on login, update with submitted time on submission
  Session data is known when someone signs in and will be used for the upcoming contact us page changes, so it makes more sense to write the session data on sign in. Writing the submitted time on submission will allow us to know if the user should be able to request the thank-you page.

jonnyshaw89 commented on app/authentication/authenticator.py:
```diff
@@ -54,21 +65,24 @@ def store_session(metadata):
 
     # also clear the secure cookie data
-    session.clear()
+    cookie_session.clear()
```
could you call `flask_logout()` here rather than deleting from db and clearing cookie?

dwyeradam replied:
This causes issues when trying to get the current_user for use in the get_answer_store in login() in session.py. We need a separate card now to ensure we're using the Flask login and logout effectively.

jonnyshaw89 commented on app/authentication/session_storage.py:
```diff
@@ -1,144 +1,112 @@
+        """
+        add submission time to session data
+        """
+        self.eq_session = self._get_user_session()
```
we shouldn't need to re-assign `self.eq_session` as it is already assigned by `self._get_user_session()`

jonnyshaw89 commented on app/authentication/session_storage.py:
```diff
@@ -1,144 +1,112 @@
-            logger.debug("eq_session_id does not exist in user's cookie")
+        eq_session = self._get_user_session()
+        if eq_session:
+            self.eq_session = None
```
I would set the None after the delete

jonnyshaw89 commented on app/globals.py:
```diff
@@ -18,6 +19,15 @@ def get_questionnaire_store(user_id, user_ik):
 
 
+def get_session_store():
+    # Sets up a single SessionStorage instance throughout app.
```
its not throughout the app its per request.

dwyeradam replied:
Lazy cut and paste from questionnaire_store above. Corrected both.

jonnyshaw89 commented on app/setup.py:
```diff
@@ -231,10 +228,13 @@ def add_blueprints(application):
     application.register_blueprint(questionnaire_blueprint)
     questionnaire_blueprint.config = application.config.copy()
 
+    application.register_blueprint(post_submission_blueprint)
```
to maintain consistency i think the import should be above each blueprint?

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -273,19 +281,11 @@ def submit_answers(routing_path, eq_id, form_type, collection_id):
-            eq_id=eq_id,
-            form_type=form_type,
-            collection_id=collection_id))
+        get_session_store().add_submission_time()
```
do we want to be more explicit here and pass the time in?

dwyeradam replied:
I didn't but go on then. Passing in datetime.utcnow().isoformat() to follow runner standards.

jonnyshaw89 commented on app/views/session.py:
```diff
@@ -87,21 +86,21 @@ def get_timeout_continue():
 
 @session_blueprint.route('/session-expired', methods=['GET'])
 def get_session_expired():
     if current_user.is_active:
```
Do we need this `if`? i think `logout_user()` checks this

ajmaddaford commented on app/authentication/session_storage.py:
```diff
@@ -1,144 +1,111 @@
-    @staticmethod
-    def store_user_id(user_id):
+    def __init__(self):
+        self.eq_session = None
```
I think we need to simplify this:
- The class should expose the properties` user_id`, `eq_session_id` and `session_data` - this will remove the need for the `get_user_id` and `get_session_data` properties. The `session_data` property should be a `SessionData` class.
- We should just need `load`, `save` and `delete` methods. To update a session we would update the `session_data` property and call save.
- The eq_session_id from the cookie should be passed in, rather than having a dependency on the flask cookie session in this class

Also, we should rename this file session_store.py and move it alongside questionnaire_store.

jonnyshaw89 replied:
i don't think we need `load`. We should do that on the init as we do with `QuestionnaireStore`

ajmaddaford commented on app/authentication/session_storage.py:
```diff
@@ -1,144 +1,111 @@
+        eq_session = self._get_user_session()
+        if eq_session and eq_session.session_data:
+            session_data = json.loads(eq_session.session_data, object_hook=lambda d: SessionData(**d))
+            session_data.add_submitted_time(submission_time)
```
Can this just be `session_data.submitted_time = submission_time`?

MIslam92 commented on app/authentication/authenticator.py:
```diff
@@ -28,14 +33,25 @@ def request_load_user(request):  # pylint: disable=unused-argument
 
 
+@user_logged_out.connect_via(ANY)
+def when_user_logged_out(sender_app, user):  # pylint: disable=unused-argument
```
Could you not silence these by just adding a _prefix before each variable passed instead wherever we pass unused arguments?

ajmaddaford replied:
I tried this and got an error when this method is called with a keyword argument that doesn't match the `_` prefixed parameters:
`when_user_logged_out() got an unexpected keyword argument 'user'`

ajmaddaford reviewed (approved):
@dwyeradam I made some changes to the pull request last week as I'm off this week. Please can you review the changes I've made.

dwyeradam reviewed (commented):
I've checked the changes Andrew made and i'm happy with the messing of my code.

dwyeradam committed:
Re-Work Session Storage

Session Store object created and global
Session Data in DB created to store previous cookie metadata
Thank you page requires login to access session data

MIslam92 reviewed (approved):
All above addressed
